services/rfq/relayer/service/handlers.go

```diff
@@ -162,6 +162,17 @@ func (q *QuoteRequestHandler) handleSeen(ctx context.Context, span trace.Span, r
 		return nil
 	}
 
+	// lock the consumed balance
+	key := getBalanceMtxKey(q.Dest.ChainID, request.Transaction.DestToken)
+	span.SetAttributes(attribute.String("balance_lock_key", key))
+	unlocker, ok := q.balanceMtx.TryLock(key)
+	if !ok {
+		// balance is locked due to concurrent request, try again later
+		span.SetAttributes(attribute.Bool("locked", true))
+		return nil
+	}
+	defer unlocker.Unlock()
```

Comment:
In the case of forwarding, won't the lock be released only after the full cycle goes through? We could safely unlock once we update status to `CommittedPending` (aka spend the destination commitable balance)

Reply:
@dwasse by "full cycle" I mean the full status path `Seen -> CommitPending -> CommittedConfirmed`. I realised that it's not actually full, but still if that's the case we're keeping the lock longer than we should've

Reply:
Yeah that makes sense, would still like to use `defer` to make this error resistant but can just add a helper function here